astroid/bases.py

```diff
@@ -372,25 +377,50 @@ def infer_call_result(self, caller, context):
         on ``object.__new__`` will be of type ``object``,
         which is incorrect for the argument in general.
         If no context is given the ``object.__new__`` call argument will
-        correctly inferred except when inside a call that requires
+        be correctly inferred except when inside a call that requires
         the additional context (such as a classmethod) of the boundnode
         to determine which class the method was called from
         """
 
-        # If we're unbound method __new__ of builtin object, the result is an
+        # If we're unbound method __new__ of a builtin, the result is an
         # instance of the class given as first argument.
-        if (
-            self._proxied.name == "__new__"
-            and self._proxied.parent.frame(future=True).qname() == "builtins.object"
-        ):
-            if caller.args:
-                node_context = context.extra_context.get(caller.args[0])
-                infer = caller.args[0].infer(context=node_context)
-            else:
-                infer = []
-            return (Instance(x) if x is not Uninferable else x for x in infer)
+        if self._proxied.name == "__new__":
+            qname = self._proxied.parent.frame(future=True).qname()
+            # Avoid checking builtins.type: _infer_type_new_call() does more validation
+            if qname.startswith("builtins.") and qname != "builtins.type":
+                return self._infer_builtin_new(caller, context)
         return self._proxied.infer_call_result(caller, context)
 
+    def _infer_builtin_new(
+        self,
+        caller: nodes.Call,
+        context: InferenceContext,
+    ) -> collections.abc.Generator[
+        nodes.Const | Instance | type[Uninferable], None, None
+    ]:
```

Comment:
Would it make sense to add the same type hints to `infer_call_result` above?

Reply:
Not that easily, because we immediately hit the main blocker for adding typing in `astroid` to my knowledge:

```shell
astroid/bases.py:388: error: "None" has no attribute "name"  [attr-defined]
astroid/bases.py:389: error: "None" has no attribute "parent"  [attr-defined]
astroid/bases.py:393: error: "None" has no attribute "infer_call_result"  [attr-defined]
```

Reply:
> Not that easily, because we immediately hit the main blocker for adding typing in `astroid` to my knowledge:

Let's leave it as is then, at least for now.